presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashBasedPartialDistinctLimit.java

```diff
@@ -0,0 +1,102 @@
+import com.facebook.presto.sql.planner.PlanVariableAllocator;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.facebook.presto.sql.relational.Expressions.call;
+import static com.facebook.presto.sql.relational.Expressions.constant;
+import static java.util.Objects.requireNonNull;
+
+public class HashBasedPartialDistinctLimit
+        implements PlanOptimizer
+{
+    private final FunctionAndTypeManager functionAndTypeManager;
+
+    public HashBasedPartialDistinctLimit(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionManager is null");
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, PlanVariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(types, "types is null");
+        requireNonNull(variableAllocator, "variableAllocator is null");
+        requireNonNull(idAllocator, "idAllocator is null");
+        if (SystemSessionProperties.isHashBasedDistinctLimitEnabled(session)) {
+            return SimplePlanRewriter.rewriteWith(new Rewriter(session, idAllocator, variableAllocator, functionAndTypeManager), plan, null);
+        }
+
+        return plan;
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final Session session;
+        private final PlanNodeIdAllocator idAllocator;
+        private final PlanVariableAllocator variableAllocator;
+        private final FunctionAndTypeManager functionAndTypeManager;
+
+        private Rewriter(Session session, PlanNodeIdAllocator idAllocator, PlanVariableAllocator variableAllocator, FunctionAndTypeManager functionAndTypeManager)
+        {
+            this.session = session;
+            this.idAllocator = idAllocator;
+            this.variableAllocator = variableAllocator;
+            this.functionAndTypeManager = functionAndTypeManager;
+        }
+
+        @Override
+        public PlanNode visitDistinctLimit(DistinctLimitNode node, RewriteContext<Void> context)
+        {
+            if (node.isPartial() && node.getLimit() <= SystemSessionProperties.getHashBasedDistinctLimitThreshold(session)) {
+                VariableReferenceExpression hashVariable;
+
+                if (node.getHashVariable().isPresent()) {
+                    hashVariable = node.getHashVariable().get();
+                }
+                else if (node.getDistinctVariables().size() == 1 && node.getDistinctVariables().get(0).getType() == BIGINT) {
+                    // see if this is a distinct on single long variable for which we don't optimize hash
```

Comment:
Is this the case? I believe we have a special optimization for hash-based aggregation on a single BIGINT key.

Reply:
Yeah that's why I use that single bigint as the variable.

Reply:
Just curious, what's the optimization? It doesn't apply to other integer types?

Reply:
hash is a long so looks like optimize for case when group by key is a single BIGINT type (which is what we use for all integer type I thought)

Reply:
we use java long for all integer SQL types. So if this is specific to the java type, it should map to TINYINT,... BIGINT.